File: 092_reverse_linked_list_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    // Time Complexity: O(n)
    // Space Complexity: O(1)
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode curr = head;
        ListNode prev = null;

        // initialize curr and prev position
        while (m > 1) {
            prev = curr;
            curr = curr.next;
            m--;
            n--;
        }

        ListNode con = prev, tail = curr;
        while (n > 0) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
            n--;
        }

        // Adjust the final connections as explained in the algorithm
        if (con != null) {
            con.next = prev;
        } else {
            head = prev;
        }

        tail.next = curr;
        return head;
    }
}

File: python/092_reverse_linked_list_II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        prev = None
        curr = head

        while m > 1:
            prev = curr
            curr = curr.next
            m -= 1
            n -= 1

        con, tail = prev, curr
        while n > 0:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
            n -= 1

        if con:
            # 1 (con) -> prev (4)
            con.next = prev
        else:
            # in:  1, 2, 3, 4, 5, m = 1, n = 4
            # out: 4, 3, 2, 1, 5
            # head originally points to 1 (from n while loop)
            # now points to 4
            head = prev

        # connect tail
        # 2 (tail) -> 5 (new curr)
        tail.next = curr
        return head
